gibbon-compiler/src/Gibbon/Passes/AddCastInstructions.hs

```diff
@@ -0,0 +1,431 @@
+                                                                                                                                        pure (insts, fcenv, fenv, fexps ++ [VarE var])
+                                                                                                                _ -> error $ "addCastsExp: Expected a variable, got " ++ show exp_ty
+                                                                                        _ -> pure (insts, fcenv, fenv, fexps ++ [VarE var]) --error $ "addCastsExp: Expected a variable, got " ++ show ty_of_var
+                                                                        _ -> pure (insts, fcenv, fenv, fexps ++ [expr]) -- error "TODO: addCastsExp: not implemented yet!!"                 
+                
+                
+                                                       ) ([], cenv, new_env, []) (zip es tys_of_es)
+             bod' <- addCastsExp fundef cenv' env' bod
+             let ex' = foldr (\ex acc -> ex acc) (LetE (v, locs, ty, ((MkProdE (L.reverse vars')))) bod') new_insts
+             pure $ ex'
+
+             
+        -- LetE (v, locs, ty, rhs@(Ext (AppE f _locs args))) bod -> do
+        --     let new_env = extendVEnv v ty env
+        --     (new_insts, cenv', env', args')
+
+        LetE (v, locs, ty, rhs@(Ext (AddCursor v' e))) bod -> do
+            let nv = case (M.lookup v cenv) of 
+                            Just v' -> v'
+                            Nothing -> v
+            let env' = extendVEnv v (CursorTy) env
+            rhs' <- addCastsExp fundef cenv env' rhs
+            bod' <- addCastsExp fundef cenv env' bod
+            pure $ (LetE (nv, locs, CursorTy, rhs')) bod'
+            -- let nv = case (M.lookup v' cenv) of 
+            --                     Just v'' -> v''
+            --                     Nothing -> v'
+            -- let nv_ty = lookupVEnv nv env
+            -- e' <- go e
+            -- bod' <- addCastsExp fundef cenv env' bod
+            -- pure $ (LetE (nv, locs, ty, rhs')) bod'
+
+            --let ty_v' = lookupVEnv v' env
+            --case (ty == ty_v') of
+            --    True -> do
+            --            e' <- go e
+            --            let env' = extendVEnv v ty env
+            --            bod' <- addCastsExp fundef cenv env' bod
+            --            pure $ (LetE (v, locs, ty, Ext (AddCursor v' e'))) bod'
+            --    False -> do 
+            --             let nv = case (M.lookup v cenv) of 
+            --                            Just v'' -> v''
+            --                            Nothing -> v
+            --             let env' = extendVEnv v ty env
+            --             rhs' <- addCastsExp fundef cenv env' rhs
+            --             bod' <- addCastsExp fundef cenv env' bod
+            --             pure $ (LetE (nv, locs, ty, rhs')) bod'
```

Comment:
there's a lot of chunks of commented code here and there. Could you clean those up? Probably just remove. We don't store old code this way since version control systems became a common tool.